File: backend/src/routes/changelog.ts
import { Hono } from 'hono';
import { getChangelog, addChangelogEntry, deleteChangelogEntry, getDraft, updateDraft, publishDraft } from '../services/changelog';

const changelog = new Hono();

// GET /api/changelog — public, anyone can view
changelog.get('/', async (c) => {
  const entries = await getChangelog();
  return c.json({ entries });
});

// POST /api/changelog — admin only
changelog.post('/', async (c) => {
  const isAuthed = c.get('auth').authenticated;
  if (!isAuthed) {
    return c.json({ error: 'Not authenticated' }, 401);
  }

  const body = await c.req.json();
  const { version, title, description, type } = body as {
    version: string;
    title: string;
    description: string;
    type: 'major' | 'minor' | 'patch';
  };

  if (!version || !title || !description || !type) {
    return c.json({ error: 'version, title, description, and type are required' }, 400);
  }

  const validTypes = ['major', 'minor', 'patch'];
  if (!validTypes.includes(type)) {
    return c.json({ error: 'type must be major, minor, or patch' }, 400);
  }

  const entry = await addChangelogEntry({ version, title, description, type });
  return c.json({ entry }, 201);
});

// DELETE /api/changelog/:version — admin only
changelog.delete('/:version', async (c) => {
  const isAuthed = c.get('auth').authenticated;
  if (!isAuthed) {
    return c.json({ error: 'Not authenticated' }, 401);
  }

  const version = c.req.param('version');
  const deleted = await deleteChangelogEntry(version);
  if (!deleted) {
    return c.json({ error: 'Version not found' }, 404);
  }
  return c.json({ success: true });
});

// ─── Draft Routes (admin only) ──────────────────────────────

// GET /api/changelog/draft — get current working draft
changelog.get('/draft', async (c) => {
  const draft = await getDraft();
  return c.json({ draft });
});

// PUT /api/changelog/draft — save draft changes
changelog.put('/draft', async (c) => {
  if (!c.get('auth').authenticated) {
    return c.json({ error: 'Not authenticated' }, 401);
  }
  const body = await c.req.json();
  const { description } = body as { description: string };
  if (description === undefined) {
    return c.json({ error: 'description is required' }, 400);
  }
  const draft = await updateDraft(description);
  return c.json({ draft });
});

// POST /api/changelog/draft/publish — publish draft as a release
changelog.post('/draft/publish', async (c) => {
  if (!c.get('auth').authenticated) {
    return c.json({ error: 'Not authenticated' }, 401);
  }
  const body = await c.req.json();
  const { version, title, type } = body as {
    version: string;
    title: string;
    type: 'major' | 'minor' | 'patch';
  };
  if (!version || !title || !type) {
    return c.json({ error: 'version, title, and type are required' }, 400);
  }
  const validTypes = ['major', 'minor', 'patch'];
  if (!validTypes.includes(type)) {
    return c.json({ error: 'type must be major, minor, or patch' }, 400);
  }
  const entry = await publishDraft(version, title, type);
  return c.json({ entry }, 201);
});

export default changelog;

File: backend/app/routes/changelog.py
"""Changelog routes — mirrors backend/src/routes/changelog.ts."""

from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..deps import admin_authenticated
from ..services.changelog import (
    add_changelog_entry,
    delete_changelog_entry,
    get_changelog,
    get_draft,
    publish_draft,
    update_draft,
)

router = APIRouter()

VALID_TYPES = ("major", "minor", "patch")


def _authed(request: Request) -> bool:
    return admin_authenticated(request)


@router.get("")
async def list_changelog(request: Request) -> dict:
    if not _authed(request):
        return JSONResponse({"error": "Not authenticated"}, status_code=401)
    return {"entries": await get_changelog()}


@router.post("")
async def create_entry(request: Request) -> dict:
    if not _authed(request):
        return JSONResponse({"error": "Not authenticated"}, status_code=401)
    body = await request.json()
    version = body.get("version")
    title = body.get("title")
    description = body.get("description")
    entry_type = body.get("type")
    if not version or not title or not description or not entry_type:
        return JSONResponse({"error": "version, title, description, and type are required"}, status_code=400)
    if entry_type not in VALID_TYPES:
        return JSONResponse({"error": "type must be major, minor, or patch"}, status_code=400)
    entry = await add_changelog_entry({"version": version, "title": title, "description": description, "type": entry_type})
    return JSONResponse({"entry": entry}, status_code=201)


@router.delete("/{version}")
async def delete_entry(version: str, request: Request) -> dict:
    if not _authed(request):
        return JSONResponse({"error": "Not authenticated"}, status_code=401)
    deleted = await delete_changelog_entry(version)
    if not deleted:
        return JSONResponse({"error": "Version not found"}, status_code=404)
    return {"success": True}


@router.get("/draft")
async def get_draft_route(request: Request) -> dict:
    if not _authed(request):
        return JSONResponse({"error": "Not authenticated"}, status_code=401)
    return {"draft": await get_draft()}


@router.put("/draft")
async def save_draft(request: Request) -> dict:
    if not _authed(request):
        return JSONResponse({"error": "Not authenticated"}, status_code=401)
    body = await request.json()
    description = body.get("description")
    if description is None:
        return JSONResponse({"error": "description is required"}, status_code=400)
    draft = await update_draft(description)
    return {"draft": draft}


@router.post("/draft/publish")
async def publish(request: Request) -> dict:
    if not _authed(request):
        return JSONResponse({"error": "Not authenticated"}, status_code=401)
    body = await request.json()
    version = body.get("version")
    title = body.get("title")
    entry_type = body.get("type")
    if not version or not title or not entry_type:
        return JSONResponse({"error": "version, title, and type are required"}, status_code=400)
    if entry_type not in VALID_TYPES:
        return JSONResponse({"error": "type must be major, minor, or patch"}, status_code=400)
    entry = await publish_draft(version, title, entry_type)
    return JSONResponse({"entry": entry}, status_code=201)